### ISSUE TITLE
NVDA doesn't read in edit field, Chrome/ Firefox, Edge

### ISSUE THREAD

**Issue by HenkMulder** [nvaccess/nvda#7801]
When entering text in an complex edit field, NVDA does not read back the text. I've tried this one in Firefox 57.0, Chrome (62.0.3202.94) and Edge.

* Browse to https://www.overleaf.com/
* Follow the link "CREATE A NEW PAPER" (to try out the editor)
* Navigate to the edit box. There should already be text in it, which you can't read. But for clarity you can delete it all (ctrl+a followed by delete), and enter some new lines of text.
* Navigate through the text with arrow keys, or try using reading commands like NVDA+up-arrow and notice that text isn't read back to you.
* If you select (parts of) the text, NVDA will read the selection.

Installed version of NVDA 2017.3
Windows 10 pro, 10.0.16299

**Comment by feerrenrut** [comment]
I can reproduce this in chrome, as described. I was not able to read the text in rich edit or "source" modes.
I confirmed that the text does not show in the accessibility viewer tool.

My guess, is this is a bug in the "overleaf" web application accessibility implementation. It would take some time to debug this to work out exactly what is going on.

I would suggest trying to contact "overleaf" to see if they are aware of the issue.

**Comment by HenkMulder** [comment]
Well I noticed this problem on more sites. However most of them you need to have an account to log in. Therefore I used overleaf as example.

The same happens on https://www.sharelatex.com, and also the editor on https://leekwars.com/ doesn't read back the text you entered.

Because the problem occurs on more places I thought it might be worth while reporting.

**Comment by feerrenrut** [comment]
Yes, it is certainly worth reporting, thank you for doing so. The problem is these are complicated examples that will be quite time consuming to investigate and find the cause of the problem.

**Comment by gdata1** [comment]
Hello. I can confirm this problem. Microsoft edge on many sites. I almost always help reload the NVDA.

**Comment by Adriani90** [comment]
I can confirm this problem still occurs in NVDA 2019.1.1, Chrome 74 and Firefox 66.0.3.
1. Open the link in the description
2. Register to the website and create a new project
3. Enter a name for the project and press ok
4. Navigate to the editors mode group with r and choose either source or ritch text
5. Press e to land on the edit field (the latex syntax 
is displayed under the edit field according to NVDA).
6. Enter text in the field or try to navigate with up and down arrows

Actual: NVDA displays the text which is entered in the edit field below the edit field element. Notice that when you exit the edit field with nvda+space bar (browse mode) and press down arrow, the text you entered is readable but NVDA is not able to read it while you are in focus mode in the edit field. The content is only reported in focus mode when you select the text with shift+arrow keys.

@jcsteh since this is impacting many students (math students and science students etc.) I think it would be worthwhile to invest some time to see if something can be done from NVDA's perspective at least for Firefox.

**Comment by feerrenrut** [comment]
A minimal test case that demonstrates the problem would go a long way towards getting this fixed.

**Comment by Adriani90** [comment]
I was able to catch an error when pressing e to navigate to the edit field in browse mode:

```
IO - inputCore.InputManager.executeGesture (19:15:22.368):
Input: kb(laptop):upArrow
ERROR - scriptHandler.executeScript (19:15:23.180):
error executing script: <bound method Gecko_ia2.script_moveByLine_back of <virtualBuffers.gecko_ia2.Gecko_ia2 object at 0x054BF410>> with gesture u'Pfeiltaste nach oben'
Traceback (most recent call last):
  File "scriptHandler.pyc", line 187, in executeScript
  File "cursorManager.pyc", line 250, in script_moveByLine_back
  File "cursorManager.pyc", line 185, in _caretMovementScriptHelper
  File "browseMode.pyc", line 1239, in _set_selection
  File "baseObject.pyc", line 31, in __get__
  File "textInfos\offsets.pyc", line 418, in _get_NVDAObjectAtStart
  File "virtualBuffers\__init__.pyc", line 186, in _getNVDAObjectFromOffset
  File "virtualBuffers\gecko_ia2.pyc", line 173, in getNVDAObjectFromIdentifier
  File "IAccessibleHandler.pyc", line 324, in normalizeIAccessible
RuntimeError: <comtypes.client.dynamic._Dispatch object at 0x054545F0> Not an IAccessible
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (19:15:23.305):
accRole failed: (-2147024809, 'Falscher Parameter.', (None, None, None, 0, None))
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (19:15:23.493):
accRole failed: (-2147024809, 'Falscher Parameter.', (None, None, None, 0, None))
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (19:15:23.927):
accRole failed: (-2147024809, 'Falscher Parameter.', (None, None, None, 0, None))
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (19:15:24.042):
accRole failed: (-2147024809, 'Falscher Parameter.', (None, None, None, 0, None))
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (19:15:24.059):
accRole failed: (-2147024809, 'Falscher Parameter.', (None, None, None, 0, None))
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (19:15:24.276):
accRole failed: (-2147024809, 'Falscher Parameter.', (None, None, None, 0, None))
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (19:15:24.295):
accRole failed: (-2147024809, 'Falscher Parameter.', (None, None, None, 0, None))
IO - inputCore.logTimeSinceInput (19:15:24.496):
2.130 sec since input
IO - speech.speak (19:15:24.496):
Speaking [u'Eingabefeld  mehrzeilig', u'\u2028\u2028']
IO - inputCore.InputManager.executeGesture (19:15:24.530):
Input: kb(laptop):downArrow
IO - inputCore.logTimeSinceInput (19:15:24.601):
0.071 sec since input
IO - speech.speak (19:15:24.601):
Speaking [u'Eingabefeld-Ende', u'HelloHellohello\\usepackage[utf8]{inputenc}']
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (19:15:24.848):
accRole failed: (-2147024809, 'Falscher Parameter.', (None, None, None, 0, None))
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (19:15:25.134):
accRole failed: (-2147024809, 'Falscher Parameter.', (None, None, None, 0, None))
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (19:15:25.144):
accRole failed: (-2147024809, 'Falscher Parameter.', (None, None, None, 0, None))
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (19:15:25.157):
accRole failed: (-2147024809, 'Falscher Parameter.', (None, None, None, 0, None))
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (19:15:25.168):
accRole failed: (-2147024809, 'Falscher Parameter.', (None, None, None, 0, None))
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (19:15:25.180):
accRole failed: (-2147024809, 'Falscher Parameter.', (None, None, None, 0, None))
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (19:15:25.191):
accRole failed: (-2147024809, 'Falscher Parameter.', (None, None, None, 0, None))
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (19:15:25.203):
accRole failed: (-2147024809, 'Falscher Parameter.', (None, None, None, 0, None))
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (19:15:25.926):
accRole failed: (-2147024809, 'Falscher Parameter.', (None, None, None, 0, None))
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (19:15:25.934):
accRole failed: (-2147024809, 'Falscher Parameter.', (None, None, None, 0, None))
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (19:15:25.943):
accRole failed: (-2147024809, 'Falscher Parameter.', (None, None, None, 0, None))
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (19:15:25.948):
accRole failed: (-2147024809, 'Falscher Parameter.', (None, None, None, 0, None))
```

**Comment by Adriani90** [comment]
As you can see, NVDA speaks the content of the editfield after I press down arrow. It says "end of edit field" although the text is visually in the edit field. However, I guess that when you type in the edit field, the text is written to that area through a different property. Unfortunately I don't understand exactly how it works. But I guess that the typing event is kind of splited from the displaying event.
I will try to build a test case.

**Comment by Adriani90** [comment]
I guess it is not really possible to create a minimal text case. i think they use a very specific role and DOMNode for the edit area. According to the accessibility checker in firefox, the role is Text Leaf and the DOMNode is textarea.ACE_text-input.

**Comment by jcsteh** [comment]
> @jcsteh since this is impacting many students (math students and science students etc.) I think it would be worthwhile to invest some time to see if something can be done from NVDA's perspective at least for Firefox.

There is almost certainly nothing that can be done here by either Firefox or NVDA. It sounds like this editor uses a custom output area, rather than using HTML contentEditable. contentEditable is currently the only way to make a web editor accessible. Similar problems exist with CodeMirror, for example. This needs to be fixed by Overleaf.

**Comment by feerrenrut** [comment]
Thanks @jcsteh, in this case I'm going to close this as can't fix. We would be happy to work with overleaf to help them fix the issue.
